gk/main.c

```diff
@@ -552,6 +552,72 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_config *gk_conf = arg;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv4
+				&& memcmp(&ltbl->tunnels[i].flow.f.v6.dst,
+				map->ip_be, sizeof(struct in_addr)) == 0) {
+			ether_addr_copy(&map->ha,
+				&ltbl->tunnels[i].nexthop_mac);
+			break;
+		}
+	}
+
+	*pcall_again = true;
+}
+
+static void
+gk_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	unsigned int i;
+	struct gk_config *gk_conf = arg;
+	struct gk_lpm *ltbl = gk_conf->lpm_tbl;
+
+	for (i = 0; i < ltbl->num_tunnels; i++) {
+		if (ltbl->tunnels[i].flow.proto == ETHER_TYPE_IPv6
```

Comment:
I think that `&ltbl->tunnels[i]` should be able to be used directly, since I don't think the `tunnels` array changes after configuration, so a pointer to the tunnel in question should stay valid.

Actually, to be really minimal, I think you could make `arg` the `&ltbl->tunnels[i].nexthop_mac`, since that's all `gk_arp_req_cb()` and `gk_nd_req_cb()` are changing. You might even be able to reduce those two functions down to one, as long as there is no work that is different between the ARP and ND versions.